### DISCUSSION TITLE
Usage in MPA frameworks with SSR?

### DISCUSSION THREAD

**Discussion by miguelrk** [maxnowack/signaldb#626] [Q&A] [answered]
**SignalDB** is just what I've been taking a go at for quite some time, but better. 🎉 Glad I stumbled with it...

The thing is, I'm currently building full-stack MPAs (rather than client-only SPAs) using the[ Deno Fresh framwork](https://fresh.deno.dev/), which is Preact-based and supports the Islands architecture, but its primarily an SSR framework. I got Fresh+SignalDB working using `signaldb-plugin-preact`, but I'm not really sure if SignalDB makes sense in a MPA architecture with traditional navigation and not client-side navigation.

The question is more of a conceptual one, rather than a how-to on integrating SignalDB with Fresh:

**Does SignalDB play well in a MPA architecture? What patterns would make sense to use SignalDB in the client-side of a MPA application with traditional (server-side) navigation?**

### Lifecycle of an optimistic UI

Considering the following diagram of the "lifecycle of an optimistic UI", I guess this pattern does make sense for MPA applications when we consider each page of the MPA to be separate, so that each optimistic UI lifecycle is tied to the page, in the sense that changing pages simply destroys everything and recreates it (wires everything together) again. Would this be wasting resources?

![0_tktJEL9mC8_RRKQh](https://github.com/maxnowack/signaldb/assets/18708856/2630a321-4512-4746-96da-4564d4ec74a8)

Any insight is greatly appreciated. Happy to have a discussion on the topic.

**Comment by maxnowack** [comment, marked as answer]
Hey @miguelrk, glad you found SignalDB!
It looks like whether SignalDB fits well with your project really depends on what you're trying to do. I'm not super familiar with MPA and the Islands architecture myself, but from what I get, SignalDB should work fine with them. Although using SignalDB might make your project a bit bigger, it could also make building your app faster and more enjoyable.

The diagram you shared comes from Meteor, and actually, when I was making SignalDB, I was thinking about how Meteor handles data. I'm aiming to do something similar where you can set up a query with SignalDB, and it takes care of fetching your data and updating your page or part of it automatically whenever your data changes or gets updated (also see #394).

But really, it depends a lot on how much data your app will be handling, how often you'll be reusing that data, and how you've set up SignalDB to save data.

**Comment by miguelrk** [comment]
Thanks for the thorough response @maxnowack! I'm glad I found about it too!

SignalDB has been working great, the only issue (it's not a SignalDB issue really) has been that data is fetched twice (once on the server and another time while hydrating the island in the client). 

I'll try to quickly explain the paper cut with SignalsDB+SSR:

In the SSR context (at least that of Deno Fresh), one cannot simply "send" reactive collection instances from the server to the client since they will not be properly serialized (functions are lost when serializing props from server to client components (or islands)). What can indeed be done, is fetching the data for the collection once on the server, and passing that data (which is indeed JSON-serializable) down to the client, where we could theoretically instantiate a reactive collection, but pass `props.data` avoiding the re-fetch on first load (during hydration). For subsequent client-side updates, the `load()` data can indeed be called again.

I guess if I can figure out a way to pass the pre-fetched `props.data` when instantiating the collection on the client, I could skip the re-fetch once on first load. That should work. Obviously when the page is reloaded, again, the server should load data, and client should use that instead of re-fetching, so I guess this solves it. I'll try it out and report back 👍🏼

**Reply by maxnowack** [reply]
You can attempt to serialize the memory adapter instance of a collection (which defaults to an array instance) if possible. Then, you could use this serialized memory adapter to instantiate the SignalDB collection. However, I suspect there might be some issues related to indexing since this approach hasn’t been tested yet. I‘ll look into this soon. If successful, this method could potentially eliminate the need for data loading on the client side entirely.

**Reply by maxnowack** [reply]
I've just added a test case, fixed the indexing and published `v0.8.9`. In this version it's possible to seed data with a memory adapter. Check out [the test case](https://github.com/maxnowack/signaldb/blob/main/packages/signaldb/__tests__/Collection.spec.ts#L471) to see how to do that 🙂

**Comment by nickchomey** [comment]
@miguelrk have you continued using signalsdb for mpa ssr? That's what I'm looking to do and am curious how it might fit that need

**Reply by miguelrk** [reply]
> @miguelrk have you continued using signalsdb for mpa ssr? That's what I'm looking to do and am curious how it might fit that need

Hey @nickchomey! 

Coincidentally, I very recently started evaluating rewriting my entire Nuxt.js (full-stack app) to be local-first, and SignalDB came to mind. Since my original post I had not given it much though, but now I can't take it out of my head 😆 .

I'm currently comparing it with other alternatives like [zerosync.dev](https://zerosync.dev/), which is also great and promising. However, I need the backend-agnostic flexibility that SignalDB offers since I'm using Cloudflare D1 (a SQLite flavor). Many other sync engines, like zerosync.dev (successor of [replicache](https://replicache.dev/)), [electricSQL](https://electric-sql.com/), [powersync](https://www.powersync.com/), etc are typically tied to PostgreSQL or MySQL, which doesn't fit my current stack.

Would love to hear your thoughts or share notes as you explore this for MPA SSR!

What would definitively help, is if @maxnowack could help us compare and contrast how signalDB defers internally to e.g. zerosync. From the surface, I think signalDB has a very lightweight and techstack-agnostic spirit both for the backend (database/API) as well as for the frontend (framework/reactivity bindings). This is great from a design perspective, but I wonder if it results in less optimization possibilities (e.g. for queries)?

Needless to say, I think local-first solutions are already fast enough, whats a couple ms more or less?... I personally would much rather have a flexible solution like signaldb, which is also very simple from its API surface (I can atest to this, as I used it before, and it worked great). The only real concern would be in this MPA use case (client-heavy, full-stack Nuxt.js app) when doing the initial load of the whole database.

**Reply by nickchomey** [reply]
Sounds like you've thought about this quite a lot! I don't really have much to say at the moment about the db, as I need to explore the docs some more and probably include signaldb in the basic proof of concept that im working on.

But I'll share what I'm thinking about. 

I'm most considering signaldb for use in a Service Worker to respond to MPA requests. As far as the main thread/dom would be concerned, responses will be coming from the backend server (and they still could, if they have a network connection and the SW decides that would be more appropriate - data or templates missing, or whatever). 

I intend for the templates to be just basic tagged template literals - either lit-html or even just hono/html. Not even web components - just html fragments that are composed together. 

On the actual backend server, Im not sure if I'd use signaldb or not. But the same template files would be there. If I do use signaldb there, I'd be inclined to have it running in-memory only, and it would load it's data at startup from a different persistent store. The idea being (if I understand it correctly) that writes will go to the actual db, then feed into signaldb through some sort of cdc mechanism, then that will automatically update the relevant templates because they are all populated with signals that update automatically when they change, which could perhaps even be cached on the server and/or in cdn. 

How this would work in the service worker, I'm not quite sure, since it is completely ephemeral. Perhaps I won't do the auto-update stuff there. Or perhaps I'll run it in a dedicated worker which can re-render fragments as-needed and cache them. But if I end up needing to just fetch data from indexeddb to render templates on each request, that's fine - it'll only be 10ms or whatever. 

I was initially (and still am) looking at just using a Signals library for this (notably alien-signals, which seems to be the most popular and promising one nowadays), but Signaldb might be a better fit because of how it has plenty more functionality and seems to be nicely agnostic to the other tooling that gets used around it. 

I doubt that I'll make use of the Optimistic UI capabilities - I'll probably just allow for the app to be read-only (but at least usable!) while offline. 

I'm also planning to use Datastar, which is a new, better version of HTMX. It uses Server Sent Events to send html fragments (normally from server, but service worker works fine) to the main thread, where they magically get merged/morphed into the DOM. It also has support for declarative Signals usage for reactivity (but the focus is on just ephemeral user interactions, rather than true state mutations - which should go to the server/db and then generate new fragments to be sent up and merged). 

Anyway, I hope this makes some sense. I'd be happy to chat more here or feel free to DM me in the Signaldb discord server!
